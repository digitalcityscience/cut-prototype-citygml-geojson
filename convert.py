import geopandas as gpd
import xml.etree.ElementTree as ET
from shapely.geometry import Polygon
import pandas as pd
import os


def parse_citygml_footprints(citygml_file_path, original_crs, target_crs='EPSG:4326'):
    tree = ET.parse(citygml_file_path)
    root = tree.getroot()

    ns = {
        'gml': 'http://www.opengis.net/gml',
        'bldg': 'http://www.opengis.net/citygml/building/1.0'
    }

    buildings = root.findall('.//bldg:Building', ns)
    print(f"found {len(buildings)} buildings")

    footprints = []
    for building in buildings:
        # Extract the ground surface polygons (or lod2TerrainIntersection)
        ground_surfaces = building.findall('.//bldg:GroundSurface//gml:Polygon', ns)

        for surface in ground_surfaces:
            exterior_coords = surface.find('.//gml:exterior//gml:LinearRing', ns)
            coords = []

            if exterior_coords is not None:
                for coord in exterior_coords.findall('./gml:pos', ns):
                    x, y = map(float, coord.text.split()[:2])  # Only take X and Y for 2D footprint
                    # coords.append((x, y))
                    coords.append((x, y))

                if coords:
                    polygon = Polygon(coords)
                    footprints.append(polygon)

    print(f"found {len(footprints)} footprints")

    return gpd.GeoDataFrame(geometry=footprints, crs=original_crs).to_crs(target_crs)


def process_directory(directory_path, original_crs, output_geojson_path, target_crs='EPSG:4326'):
    all_footprints = []

    for filename in os.listdir(directory_path):
        if filename.endswith('.gml') or filename.endswith('.xml'):
            file_path = os.path.join(directory_path, filename)
            gdf = parse_citygml_footprints(file_path, original_crs, target_crs)
            all_footprints.append(gdf)

    combined_gdf = gpd.GeoDataFrame(pd.concat(all_footprints, ignore_index=True), crs=target_crs)
    combined_gdf.to_file(output_geojson_path, driver='GeoJSON')


def write_to_geojson(gdf, output_file_path):
    gdf.to_file(output_file_path, driver='GeoJSON')


def main():
    output_geojson_path = './data/out/footprints.geojson'
    process_directory("./data/LoD2_CityGML_HH_2016", 'EPSG:25832', output_geojson_path)


if __name__ == "__main__":
    main()
