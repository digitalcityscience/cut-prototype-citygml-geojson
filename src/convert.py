import geopandas as gpd
import xml.etree.ElementTree as ET
from shapely.geometry import Polygon
import pandas as pd
import os
from pathlib import Path

ns = {
    'gml': 'http://www.opengis.net/gml',
    'bldg': 'http://www.opengis.net/citygml/building/1.0',
    'gen': 'http://www.opengis.net/citygml/generics/1.0',
}

target_crs = 'EPSG:4326'
original_crs = 'EPSG:25832'


def get_attrib(path, building):
    element = building.find(path, ns)
    return element.text if element is not None else None


def get_ground_attributes(building):
    return {
        "lage": get_attrib('.//gen:stringAttribute[@name="DatenquelleLage"]/gen:value', building),
        "bodenhoehe": get_attrib('.//gen:stringAttribute[@name="DatenquelleBodenhoehe"]/gen:value', building),
    }


def get_roof_attributes(building):
    return {
        "dachhoehe": get_attrib('.//gen:stringAttribute[@name="DatenquelleDachhoehe"]/gen:value', building),
        "measuredHeight": get_attrib('bldg:measuredHeight', building)
    }


def extract_ground_attributes(polygon, building):
    id = polygon.attrib["{http://www.opengis.net/gml}id"]
    attributes = get_ground_attributes(building)
    attributes["id"] = id
    attributes["surface_type"] = "ground"
    return attributes


def extract_roof_attributes(polygon, building):
    id = polygon.attrib["{http://www.opengis.net/gml}id"]
    attributes = get_roof_attributes(building)
    attributes["surface_type"] = "roof"
    attributes["id"] = id
    return attributes


def extract_geometry(polygon):
    exterior_coords = polygon.find('.//gml:exterior//gml:LinearRing', ns)
    assert exterior_coords is not None
    coords = [(
        float(coord.text.split()[0]),
        float(coord.text.split()[1]),
    ) for coord in exterior_coords.findall('./gml:pos', ns)]
    return Polygon(coords)


def extract_surfaces(citygml_file_path, node, extract_attributes):
    tree = ET.parse(citygml_file_path)
    root = tree.getroot()
    buildings = root.findall('.//bldg:Building', ns)

    all_features, all_attributes = [], []
    for building in buildings:
        for polygon in building.findall(f'.//bldg:{node}//gml:Polygon', ns):
            attributes = extract_attributes(polygon, building)
            geometry = extract_geometry(polygon)
            all_attributes.append(attributes)
            all_features.append(geometry)

    gdf = gpd.GeoDataFrame(geometry=all_features, crs=original_crs).assign(**pd.DataFrame(all_attributes))
    gdf.set_index('id', inplace=True)
    return gdf.to_crs(target_crs)


def extract_all_surface_types(directory_path):
    ground_surfaces = []
    roof_surfaces = []
    files = os.listdir(directory_path)
    for filename in filter(lambda f: f.endswith('.gml') or f.endswith('.xml'), files):
        print(f"processing {filename}")
        path = os.path.join(directory_path, filename)

        grounds = extract_surfaces(path, "GroundSurface", extract_ground_attributes)
        ground_surfaces.append(grounds)

        roofs = extract_surfaces(path, "RoofSurface", extract_roof_attributes)
        roof_surfaces.append(roofs)

    return dict(
        ground=gpd.GeoDataFrame(pd.concat(ground_surfaces), crs=target_crs),
        roof=gpd.GeoDataFrame(pd.concat(roof_surfaces), crs=target_crs)
    )


def convert(directory_path, output_geojson_path):
    Path(output_geojson_path).mkdir(parents=True, exist_ok=True)
    ret = extract_all_surface_types(directory_path)
    for name, df in ret.items():
        outfile = os.path.join(output_geojson_path, f"{name}.geojson")
        df.to_file(outfile, driver='GeoJSON')
