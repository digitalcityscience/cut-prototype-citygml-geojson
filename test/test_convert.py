import os
import pathmagic
import tempfile
from convert import extract_all_surface_types, convert

assert pathmagic


def test_extract_all_surface_types(snapshot):
    df = extract_all_surface_types("./")
    assert df["ground"].to_json() == snapshot
    assert df["roof"].to_json() == snapshot


def test_convert(snapshot):
    tmp = tempfile.TemporaryDirectory()
    dir = tmp.name
    convert("./", dir)
    assert open(os.path.join(dir, "ground.geojson")).read() == snapshot
    assert open(os.path.join(dir, "roof.geojson")).read() == snapshot
