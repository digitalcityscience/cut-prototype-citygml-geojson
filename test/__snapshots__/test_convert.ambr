# serializer version: 1
# name: test_convert
  '''
  {
  "type": "FeatureCollection",
  "crs": { "type": "name", "properties": { "name": "urn:ogc:def:crs:OGC:1.3:CRS84" } },
  "features": [
  { "type": "Feature", "properties": { "id": "DEHH_00000001-0000-2001-b538-8763db60db60", "lage": "1000", "bodenhoehe": "1300", "surface_type": "ground" }, "geometry": { "type": "Polygon", "coordinates": [ [ [ 8.495865635956811, 53.915932415761503 ], [ 8.495879797235796, 53.915956275830453 ], [ 8.495949122771419, 53.91594192513098 ], [ 8.49593494623328, 53.915918065006167 ], [ 8.495865635956811, 53.915932415761503 ] ] ] } },
  { "type": "Feature", "properties": { "id": "DEHH_00000002-0000-2001-b538-8763db60db60", "lage": "1000", "bodenhoehe": "1300", "surface_type": "ground" }, "geometry": { "type": "Polygon", "coordinates": [ [ [ 8.489742551194936, 53.914275016543385 ], [ 8.489445805017125, 53.914289098077063 ], [ 8.489461466742043, 53.914404284953463 ], [ 8.489758213622176, 53.914390212370876 ], [ 8.489742551194936, 53.914275016543385 ] ] ] } }
  ]
  }
  
  '''
# ---
# name: test_convert.1
  '''
  {
  "type": "FeatureCollection",
  "crs": { "type": "name", "properties": { "name": "urn:ogc:def:crs:OGC:1.3:CRS84" } },
  "features": [
  { "type": "Feature", "properties": { "id": "DEHH_00000001-0007-2000-b538-8763db60db60", "dachhoehe": "5000", "measuredHeight": "3.726", "surface_type": "roof" }, "geometry": { "type": "Polygon", "coordinates": [ [ [ 8.495879797235796, 53.915956275830453 ], [ 8.495865635956811, 53.915932415761503 ], [ 8.495900298659025, 53.915925244914654 ], [ 8.495914459955397, 53.915949104979575 ], [ 8.495879797235796, 53.915956275830453 ] ] ] } },
  { "type": "Feature", "properties": { "id": "DEHH_00000001-0008-2000-b538-8763db60db60", "dachhoehe": "5000", "measuredHeight": "3.726", "surface_type": "roof" }, "geometry": { "type": "Polygon", "coordinates": [ [ [ 8.49593494623328, 53.915918065006167 ], [ 8.495949122771419, 53.91594192513098 ], [ 8.495914459955397, 53.915949104979575 ], [ 8.495900298659025, 53.915925244914654 ], [ 8.49593494623328, 53.915918065006167 ] ] ] } },
  { "type": "Feature", "properties": { "id": "DEHH_00000002-0005-2000-b538-8763db60db60", "dachhoehe": "5000", "measuredHeight": "8.968", "surface_type": "roof" }, "geometry": { "type": "Polygon", "coordinates": [ [ [ 8.489461466742043, 53.914404284953463 ], [ 8.489445805017125, 53.914289098077063 ], [ 8.489501826712122, 53.914344406666807 ], [ 8.489461466742043, 53.914404284953463 ] ] ] } },
  { "type": "Feature", "properties": { "id": "DEHH_00000002-0007-2000-b538-8763db60db60", "dachhoehe": "5000", "measuredHeight": "8.968", "surface_type": "roof" }, "geometry": { "type": "Polygon", "coordinates": [ [ [ 8.489445805017125, 53.914289098077063 ], [ 8.489742551194936, 53.914275016543385 ], [ 8.489750389955146, 53.914332618983828 ], [ 8.489501826712122, 53.914344406666807 ], [ 8.489445805017125, 53.914289098077063 ] ] ] } },
  { "type": "Feature", "properties": { "id": "DEHH_00000002-0008-2000-b538-8763db60db60", "dachhoehe": "5000", "measuredHeight": "8.968", "surface_type": "roof" }, "geometry": { "type": "Polygon", "coordinates": [ [ [ 8.489758213622176, 53.914390212370876 ], [ 8.489461466742043, 53.914404284953463 ], [ 8.489501826712122, 53.914344406666807 ], [ 8.489750389955146, 53.914332618983828 ], [ 8.489758213622176, 53.914390212370876 ] ] ] } }
  ]
  }
  
  '''
# ---
# name: test_extract_all_surface_types
  '{"type": "FeatureCollection", "features": [{"id": "DEHH_00000001-0000-2001-b538-8763db60db60", "type": "Feature", "properties": {"lage": "1000", "bodenhoehe": "1300", "surface_type": "ground"}, "geometry": {"type": "Polygon", "coordinates": [[[8.495865635956811, 53.9159324157615], [8.495879797235796, 53.91595627583045], [8.49594912277142, 53.91594192513098], [8.49593494623328, 53.91591806500617], [8.495865635956811, 53.9159324157615]]]}}, {"id": "DEHH_00000002-0000-2001-b538-8763db60db60", "type": "Feature", "properties": {"lage": "1000", "bodenhoehe": "1300", "surface_type": "ground"}, "geometry": {"type": "Polygon", "coordinates": [[[8.489742551194936, 53.914275016543385], [8.489445805017125, 53.91428909807706], [8.489461466742043, 53.91440428495346], [8.489758213622176, 53.914390212370876], [8.489742551194936, 53.914275016543385]]]}}]}'
# ---
# name: test_extract_all_surface_types.1
  '{"type": "FeatureCollection", "features": [{"id": "DEHH_00000001-0007-2000-b538-8763db60db60", "type": "Feature", "properties": {"dachhoehe": "5000", "measuredHeight": "3.726", "surface_type": "roof"}, "geometry": {"type": "Polygon", "coordinates": [[[8.495879797235796, 53.91595627583045], [8.495865635956811, 53.9159324157615], [8.495900298659025, 53.91592524491465], [8.495914459955397, 53.915949104979575], [8.495879797235796, 53.91595627583045]]]}}, {"id": "DEHH_00000001-0008-2000-b538-8763db60db60", "type": "Feature", "properties": {"dachhoehe": "5000", "measuredHeight": "3.726", "surface_type": "roof"}, "geometry": {"type": "Polygon", "coordinates": [[[8.49593494623328, 53.91591806500617], [8.49594912277142, 53.91594192513098], [8.495914459955397, 53.915949104979575], [8.495900298659025, 53.91592524491465], [8.49593494623328, 53.91591806500617]]]}}, {"id": "DEHH_00000002-0005-2000-b538-8763db60db60", "type": "Feature", "properties": {"dachhoehe": "5000", "measuredHeight": "8.968", "surface_type": "roof"}, "geometry": {"type": "Polygon", "coordinates": [[[8.489461466742043, 53.91440428495346], [8.489445805017125, 53.91428909807706], [8.489501826712122, 53.91434440666681], [8.489461466742043, 53.91440428495346]]]}}, {"id": "DEHH_00000002-0007-2000-b538-8763db60db60", "type": "Feature", "properties": {"dachhoehe": "5000", "measuredHeight": "8.968", "surface_type": "roof"}, "geometry": {"type": "Polygon", "coordinates": [[[8.489445805017125, 53.91428909807706], [8.489742551194936, 53.914275016543385], [8.489750389955146, 53.91433261898383], [8.489501826712122, 53.91434440666681], [8.489445805017125, 53.91428909807706]]]}}, {"id": "DEHH_00000002-0008-2000-b538-8763db60db60", "type": "Feature", "properties": {"dachhoehe": "5000", "measuredHeight": "8.968", "surface_type": "roof"}, "geometry": {"type": "Polygon", "coordinates": [[[8.489758213622176, 53.914390212370876], [8.489461466742043, 53.91440428495346], [8.489501826712122, 53.91434440666681], [8.489750389955146, 53.91433261898383], [8.489758213622176, 53.914390212370876]]]}}]}'
# ---
